subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/ReservationClientFactory.kt

```diff
@@ -0,0 +1,92 @@
+    fun create(
+        configuration: InstrumentationConfiguration.Data,
+        executionParameters: ExecutionParameters,
+        testsToRun: List<TestWithTarget>
+    ): ReservationClient
+
+    class Impl(
+        private val logger: CILogger,
+        private val buildId: String,
+        private val buildType: String,
+        private val projectName: String,
+        private val kubernetesCredentials: KubernetesCredentials,
+        private val registry: String,
+        private val output: File,
+        private val logcatDir: File
+    ) : ReservationClientFactory {
+
+        override fun create(
+            configuration: InstrumentationConfiguration.Data,
+            executionParameters: ExecutionParameters,
+            testsToRun: List<TestWithTarget>
+        ): ReservationClient {
+            val emulatorsLogsReporter = EmulatorsLogsReporter(
+                outputFolder = output,
+                logcatTags = executionParameters.logcatTags,
+                logcatDir = logcatDir
+            )
+            val androidDebugBridge = AndroidDebugBridge(
+                logger = { logger.info(it) }
+            )
+            return if (isLocalRun(testsToRun)) {
+                LocalReservationClient(
+                    androidDebugBridge = androidDebugBridge,
+                    devicesManager = AdbDevicesManager(logger = object : Logger {
+                        override fun notify(message: String, error: Throwable?) {
+                            logger.critical(message, error)
+                        }
+
+                        override fun log(message: String) {
+                            logger.info(message)
+                        }
+                    }),
+                    configurationName = configuration.name,
+                    logger = logger,
+                    emulatorsLogsReporter = emulatorsLogsReporter
+                )
+            } else {
+                KubernetesReservationClient(
+                    androidDebugBridge = androidDebugBridge,
+                    kubernetesClient = DefaultKubernetesClient(
+                        ConfigBuilder()
+                            .withCaCertData(kubernetesCredentials.caCertData)
+                            .withMasterUrl(kubernetesCredentials.url)
+                            .withOauthToken(kubernetesCredentials.token)
+                            .build()
+                    )
+                        .inNamespace(executionParameters.namespace),
+                    configurationName = configuration.name,
+                    projectName = projectName,
+                    logger = logger,
+                    buildId = buildId,
+                    buildType = buildType,
+                    emulatorsLogsReporter = emulatorsLogsReporter,
+                    registry = registry
+                )
+            }
+        }
+
+        private fun isLocalRun(testsToRun: List<TestWithTarget>): Boolean {
```

Comment:
Duplication is an indicator of poor architecture. Add comment for future rewriting

Reply:
> Duplication is an indicator of poor architecture

It depends.

Reply:
done